torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1170,14 +1426,76 @@ def _register_param_handle(self, handle: FlatParamHandle) -> None:
         if handle not in self._handles:
             self._handles.append(handle)
 
+    @torch.no_grad()
+    def _unshard(
+        self,
+        handles: List[FlatParamHandle],
+    ) -> None:
+        """
+        Unshards the handles in ``handles``. If the handles are in
+        :meth:`summon_full_params` and are using mixed precision, then they are
+        forced to full precision.
+
+        Postcondition: Each handle's ``FlatParameter`` 's data is the padded
+        unsharded flattened parameter on the compute device.
+        """
+        for handle in handles:
+            with torch.cuda.stream(self._streams["pre_all_gather"]):
+                ran_pre_unshard = handle.pre_unshard()
+                if ran_pre_unshard:
+                    self._streams["all_gather"].wait_stream(self._streams["pre_all_gather"])
```

Comment:
Only if the pre-unshard was not a no-op do we need to make this `wait_stream()` call.